pandas/core/dtypes/cast.py

```diff
@@ -1611,13 +1611,42 @@ def cast_scalar_to_array(
     ndarray of shape, filled with value, of specified / inferred dtype
 
     """
+    # that's what the type annotation indicates
+    assert isinstance(dtype, (type(None), str, np.dtype))
+
     if dtype is None:
-        dtype, fill_value = infer_dtype_from_scalar(value)
+        dtype, value = infer_dtype_from_scalar(value)
     else:
-        fill_value = value
+        if not isinstance(dtype, np.dtype):
+            dtype = np.dtype(dtype)
+        empty = shape and not any(shape)
+        # dtype coercion when empty: sometimes yes, sometimes no?
+
+        if not empty and is_integer_dtype(dtype) and isna(value):
+            # coerce if we have nan for an integer dtype
+            dtype = np.dtype("float64")
+        elif isinstance(dtype, np.dtype) and dtype.kind in ("U", "S"):
+            # we need to coerce to object dtype to avoid
+            # to allow numpy to take our string as a scalar value
+            dtype = np.dtype("object")
+            if not isna(value):
+                value = ensure_str(value)
+        elif dtype.kind == "m":
```

Comment:
could share the "m" and "M" code and for the is_valid_nat_for_dtype case use `value = dtype.type("NaT", "ns")`